### ISSUE TITLE
Forcefuly redrawing exactly one component (and children)

### ISSUE THREAD

**Issue by IngwiePhoenix** [preactjs/preact#1674]
Okay, to be fair: I need to excuse myself for this probably incorrect title, but I had no idea what terms to use instead. English is not exactly my primary language :)

So, I am currently working on a project which I would like to render mostly on the server in order to achieve:
* SEO across multiple search engines
* Lowest possible "time to/until content" (TTC)

Now usually, this would mean that one would basically write one application that runs on the browser, and reuse the exact same logic on the server to replicate said browser logic, but have it render on a server instead and send that result to the browser for it to hydrate the sent HTML into an interactive application.

But, what if the server's routing logic is actually different than to what would be used on the client? Well - that is very much my big issue. The server's request processing structure is much, much different than what `preact-router` would do.

In my case, I am using NestJS, a TypeScript based MVC framework. Basically, imagine a TypeScript version of Laravel or Yii (both PHP). Without going in-depth, you can see that the routing logic is going to be different. And this is very much where my problem comes from.

Basically, I want to make the following possible:
* Have a multi-page application,
* that loads a different component per page,
* but keeps key components in place.

Imagine something like this:

```html
<html>
  <head>
    <!-- ... probably using preact-helmet here. -->
  </head>
  <body>
    <MainMenuComponent />
    <UserMenuComponent />
    <CurrentPageComponent /> <!-- This is the important one! -->
    <FooterComponent />
  </body>
</html>
```

So far, I have come across the idea that I would implement a custom `View` class (`Presenter` would work too, probably) which would take in a component, the arguments (properties) passed to it, and possibly some other data (like a base state, for instance? Not that important for now, though). This class then decides how to return the response - as a full HTML body or a JSON object containing **only** the contents of the actual current page. In addition, that JSON object would contain a reference to the component that the client should load and render ontop of that HTML. Something like this:

```json
{
  "view": "home.sendFeedback",
  "html": "..."
}
```

The client would insert the pre-rendered HTML in place of `CurrentPageComponent`, and tell load the script referenced in `view`. This script would essentially return a component that would be given to Preact to render.

But - how do I tell Preact to drop the `CurrentPageComponent` and replace it with the component that had just been downloaded? My initial idea was to create a hierarchy of applications:

```js
import {Menu, User, Footer} from "basePage";

// Managed by webpack. Contains a list of all the views that could be rendered.
// Uses import(module:string):object to load asynchronously.
// Each view in the object is essentially a function returning the promise given from import()
import Views from "viewCollection";

var currentPageId = JSON.parse(document.getElementById("currentId").innerHTML);
var currentPage = Views[currentPageId.id];

import {render} from "preact";

render(Menu, document.getElementById("menu"));
render(User, document.getElementById("user"));
render(Footer, document.getElementById("footer"));

render(currentPage, document.getElementById("main"));
```

Now, whenever the user navigates to a link, that would in theory navigate the page, I could use something like the History API instead to pick up on the click and instead send a request to the server to please only return me the view and pre-rendered HTML data - then load the component and tell Preact to drop what currently is in `#main`, replace it with the new HTML and have it re-render the new component over it (or, hydrate it actually).

Is this a usable aproach? Is it okay to have multiple `render()` calls like this? Or will this have some side effects?

Now I know that my explanation skills aren't exactly golden - so I really hope I was able to express and explain what I want to achieve. I come from a lot of PHP development and really like the workflow of rendering on the server and sending that to the client to reduce the TTC, since that also has the positive side-effect of reducing the overall amount of data that the client needs to receive. In some case, that actually might be no additional javascript at all. When displaying stuff like pre-rendered Markdown pages, there is no need to attach additional JavaScript to them at all, so it really would be like when AJAX was used to (and actually still is) dynamically update a currently seen page with new content.

Hopefully I can find some help here. :)

Kind regards,
Ingwie

**Comment by pmkroeker** [comment]
[This react article](https://dev.to/marvelouswololo/how-to-server-side-render-react-hydrate-it-on-the-client-and-combine-client-and-server-routes-1a3p) might point you in a good direction. It does use Express instead of Nest, but the basic idea is the same. Unfortunately there are no links to headers but the section under `Combine the Express router with React Router` (~ half way down) is a good starting place.

Again this uses a certain server-side routing structure than points all sub-routes through the same endpoint (which sounds like it might not be what you have).

TLDR: The general use for SSR + (p)react-router is to point each sub-route through the same handler and allow the app to properly render what it should. The router will render the proper route on the server and send it, and will hydrate when it arrives.

**Comment by IngwiePhoenix** [comment]
That TL;DR got me into reading the article - well, I opened it and i will read it next. Thank you very much for it! Let's see if it helps :)

**Comment by IngwiePhoenix** [comment]
Yeah, it is close, but not what I was looking for. But ut did give me an idea or two on how to implement what I am looking for.

But it did not answer my base question: Is it safe/okay to have multiple `render()` calls on different elements and one specificaly for `hydrate()`?

**Comment by marvinhagemeister** [comment]
@IngwiePhoenix yup, that's perfectly fine and there are a lot of teams using Preact that way :+1: Each tree is separate and they only thing that's shared internally is the queue which processes updates :100:
